### DISCUSSION TITLE
Version 2.4 Beta 1

### DISCUSSION THREAD

**Discussion by vkbo** [vkbo/novelWriter#1779] [General]
### Release Notes

This is a beta release of the next release version, and is intended for testing purposes. Please be careful when using this version on live writing projects, and make sure you take frequent backups.

Please check the changelog for an overview of changes. The full release notes will be added to the final release.

### Detailed Changelog

**Major Features**

* A global search feature has been added to the main sidebar. The search panel replaces the project and novel tree when activated. PR #1775. Issue #894.
* A new shortcode to highlight text has been added. PR #1715. Issue #705.
* A new heading format for hard scene breaks has been added. It uses `###!` heading markup. The only affect this has in the GUI is that these headings can be independently formatted in the Manuscript tool. PR #1753. Issue #1050.
* The document editor and viewer now have a dropdown menu in the header listing all headings of the current document for quick navigation. The list is capped at 30 entries. PR #1764. issue #1059.

**Build Tool Improvements**

* The Manuscript Build Tool now has a word stats section below the preview that shows a number of word and character counts for the previewed text. PR #1717. Issues #1114 and #1116.
* The Manuscript Build Tool now shows an outline of the previewed document as a tab next to the build settings list. PR #1768. Issue #1765.
* Tabs handling in HTML output now has a separate setting from the other format. PR #1723.
* Hard line breaks can now be excluded from Markdown builds. PR #1723. Issue #944.
* It is now possible to control the centring and page breaks of partition, chapter and scene headings. PR #1723. Issues #1117 and #1661.
* Special titles (`#!`) can now be used in notes as well. PR #1723.
* Meta data categories can be filtered out from the manuscript. PR #1723. issue #1132.
* Any heading in a novel document can now be hidden in the manuscript. PR #1759. Issue #1756.
* First line of a paragraph can now be indented in the manuscript. PR #1761. Issue #906.
* Each meta data entry in HTML builds have a new class assigned to it that matches the tag used in the text. PR #1767. Issue #1134.

**Other Improvements**

* The percentage progress counter in the editor document footer now counts progress per character instead of per line. This is only noticeable on short documents. PR #1725.
* Some improvements have been made to terms on the GUI and some strings have been simplified in order to be easier to understand and to translate. PR #1727. Issue #1726.
* Dates are now formatted according to the selected locale, if such a locale is available. If not, it falls back to the local system locale. PR #1755. Issue #1739.

**Code Improvements**

* The tokenization of the novelWriter markup format has been refactored and improved. PR #1724.
* A way to read project documents fast has been added. It is useful many places in the code where only the text is needed, not the meta data. PR #1777.

<hr /><em>This discussion was created from the release <a href='https://github.com/vkbo/novelWriter/releases/tag/v2.4b1'>Version 2.4 Beta 1</a>.</em>

**Comment by awqk** [comment]
Thanks for all your work creating this update!

I have not found any issues, but maybe you  can add a few pixels of default window height to fit this panel:
![gnome-shell-screenshot-5luzi0](https://github.com/vkbo/novelWriter/assets/160403624/5bfdbc3c-cac1-4315-9f5a-81814526d09a)

**Reply by vkbo** [reply]
You can just resize the window to whatever size you want. The new layout is flexible to allow for different sizes of text strings for different languages, so there isn't a size that fits all panels by default.

**Reply by awqk** [reply]
I am aware of this, but the new size doesn't persist.

I remember querying the toolkit (not Qt) about the render size of all panels to choose the window size accordingly. This wouldn't work well in this case because the Format Panel runs longer than the others, but you get the idea.

**Reply by vkbo** [reply]
The problem is that the size is determined also by the window manager. Calculating the correct size is non-trivial, and the Qt framework doesn't auto-resize when the container is scrollable, so what I did is make a best guess at a sensible default size. From the screenshot it looks like only a few pixels are missing, but that's not the case on my desktop. It also depends on choice of GUI font and size, so it is much cleaner to just keep everything dynamic. It certainly causes a lot less problems over all.

Anyway, the size should absolutely persist. But it persists per project for project-specific tools and dialogs. Most GUI settings fall in this category.

**Reply by vkbo** [reply]
Inside the project's storage folder there is a file `meta/options.json` that holds all the last state information of all the project-specific GUI components. For this dialog, it should be in:

```json
"GuiBuildSettings": {
  "treeWidth": 324,
  "filterWidth": 232,
  "winWidth": 764,
  "winHeight": 602
},
```

You can check there that it persists.

**Reply by awqk** [reply]
Agreed, it's a per-project setting, which had me confused.

**Comment by vkbo** [comment]
Just a heads up. There is one issue I've identified in the beta release. If you make a search in the new project search panel, and edit one of the documents in the result, the search does not update to reflect that, so all results will be off the mark if you added text before their position. This has been fixed for the upcoming release candidate.

**Reply by awqk** [reply]
Also: Ctrl+Shift+F does not pick up selected text.

**Reply by vkbo** [reply]
It does not. The shortcut was intended to just switch pane. Since these panels are completely independent of each other, unlike the search tool in the editor, they don't actually talk to each other. The editor doesn't even need to be open. So it would require logic with a lot of conditions, and for the search panel to reach into the editor, to get that to work.

I can look into it though.

**Reply by awqk** [reply]
With global search, I usually want "Search Next File" turned off in the editor. It looks as if the 'off' setting does not persist.

**Reply by vkbo** [reply]
It saves the state when the search box is closed, so if the box disappears by other means, it isn't saved. This whole part of the code is really old and doesn't follow the updated closing process that most of the rest of the code does, so it's possible it misses saving the state in some cases.

Search setting are part of the main GUI settings and are saved to the app config file.

I can rewrite the code to store the state of the switch in the config class on each toggle, like is the standard in the code now.

**Reply by vkbo** [reply]
Fixed in #1795

The search box was caching the button states in the widget class, and only forwarding them to the config class on close as mentioned. I rewrote it so they are changed directly in config when you toggle the buttons, and read from there as well when you perform search tasks. Therefore the config will always be in the correct state.

**Comment by awqk** [comment]
I like the new wording for "Replace Straight Double Quotes" and "Replace Straight Single Quotes". The earlier iterations had me wondering.

I am still not sure about "Wrap Double Quotes". Is this supposed to add quotation marks to a selection? How about simply writing "Quote (Double Quotes)" and "Quote (Single Quotes)"? If this feature exists, I also wonder about the inverse operation "Unquote".

How about adding some smartness to this:

“quote me --> “quote me” and not ““quote me”

**Reply by vkbo** [reply]
> I am still not sure about "Wrap Double Quotes". Is this supposed to add quotation marks to a selection? 

Yes. It's intended as a way to apply the selected quote style without having to jump through hoops with awkward keyboard codes to type the quote symbols in case the auto-replace feature makes the wrong decision, or is turned off.

**Reply by awqk** [reply]
I can think of using this while editing, for example to insert something within a range of direct speech:

`“The waitress might know, don't you think? I think I should ask her.”`

edit -->

`“The waitress might know, don't you think? He paused to pick his nose. I think I should ask her.”`

2x Wrap Double Quotes -->

`““The waitress might know, don't you think?” He paused to pick his nose. “I think I should ask her.””`

Wouldn't it be better to get this result? -->

`“The waitress might know, don't you think?” He paused to pick his nose. “I think I should ask her.”`

This could be related to the new epic #1773.
